AS Structured Events Guide [ci skip] (#56156)
rails/rails: guides/source/active_support_structured_events.md

**DO NOT READ THIS FILE ON GITHUB, GUIDES ARE PUBLISHED ON <https://guides.rubyonrails.org>.**

Active Support Structured Events
================================

Active Support provides a unified API for reporting *structured events* inside your application. A structured event is a

> named record of something that occured in your application, including what happened, when, and from where. They form the

Reviewer: ```suggestion
named record of something that occurred in your application, including what happened, when, and from where. They form the
```

Rails.event.notify("order_#{order.id}", {})  # String interpolation creates high cardinality
```

Names should reflect *what happened*, not how or why. Variable data belongs in the payload, not the event name.
Event names can be reused, but repeated events should represent the same "thing" conceptually.

### Event Payloads

Payloads are Ruby hashes containing event-specific attributes. Consider an order fulfillment workflow:

```ruby
# When an order is created
Rails.event.notify("order_created",
  order_id: order.id,
  shop_id: shop.id,
  total_amount: order.total_price,
  currency: order.currency,
  line_item_count: order.line_items.count
)

# When payment is processed
Rails.event.notify("payment_processed",
  order_id: order.id,
  payment_id: payment.id,
  amount: payment.amount,
  currency: payment.currency,
  gateway: payment.gateway_name,
  status: "captured"
)

# When shipping label is created
Rails.event.notify("shipping_label_generated",
  order_id: order.id,
  fulfillment_id: fulfillment.id,
  carrier: "USPS",
  tracking_number: fulfillment.tracking_number,
  shipping_method: "standard"
)
```

The resulting event includes automatically-attached metadata:

```ruby

> # Emits event:

Reviewer: Can we make this match one of the 3 events above?

* Keep payloads focused, and only include relevant data
* Omit nil values
* Follow semantic conventions when available (e.g., OpenTelemetry)

Further details on object payloads and automatic key normalization can be found in the [Event Objects][] section of the API documentation.

### Debug Events

Debug events are emitted only when debug mode is enabled:

```ruby
# High-volume diagnostic information
Rails.event.debug("inventory_check_performed",
  order_id: order.id,
  sku: line_item.sku,
  available_quantity: inventory.quantity
)

# Expensive payload: only computed in debug mode
Rails.event.debug("large_data_exported",
  user_id: user.id,
  export_id: export.id
) do
  {
    exported_data: generate_large_export_data, # This could be computationally intensive
    record_count: export.items.count,
    export_time_ms: export.elapsed_time
  }
end
```

Debug events help diagnose issues in development or when explicitly enabled via `Rails.event.with_debug`.
For example:

```ruby
class ApplicationController < ActionController::Base
  around_action :with_event_debug_if_requested

  private
    def with_event_debug_if_requested
      if params[:debug]

>         Rails.event.with_debug(&block)

Reviewer: ```suggestion
        Rails.event.with_debug { yield }
```

This prevents high-volume diagnostic telemetry from flooding production logs by default.

Further details on debug events can be found in the [Debug Events][] section of the API documentation.

Subscribers
-----------

A subscriber receives events and is responsible for exporting them to a destination such as:

* a log file
* a streaming platform
* a metrics or analytics system
* a SaaS observability backend

Here's an example of a subscriber that exports order events to an analytics platform:

```ruby
class OrderAnalyticsSubscriber
  def emit(event)
    case event[:name]
    when "order_created", "order_completed", "payment_processed"
      AnalyticsPlatform.track(
        event_name: event[:name],
        timestamp: event[:timestamp],
        properties: event[:payload],
        source: event[:source_location]
      )
    end
  end
end

Rails.event.subscribe(OrderAnalyticsSubscriber.new)
```

Subscribers can also filter events to receive only specific types:

```ruby

> Rails.event.subscribe(
>   OrderAnalyticsSubscriber.new,
>   filter: ->(event) { event[:name].start_with?("order_") }
> )

Reviewer: The filter needs to be passed as a block:
```suggestion
Rails.event.subscribe(OrderAnalyticsSubscriber.new) { |event|
  event[:name].start_with?("order_")
}
```

```

Real-world subscriber implementations often:

* Batch events for efficiency
* Handle errors gracefully to avoid disrupting the application
* Perform async export to external systems
* Apply additional filtering or transformation

For more information on [Subscribers][] and [Filtered Subscriptions][], please see their respective sections within the Structured Events API documentation.

Tags
----

Tags annotate events with domain-specific labels that help categorize and filter events across your application. Use tags when you want to mark related events with categorical attributes without repeating them in every payload.

> Here's an example tagging payment events by gateway and environment:

Reviewer: ```suggestion
Here's an example tagging payment events by gateway:
```
(Or we should update the example to include an env tag)

[Subscribers]: https://api.rubyonrails.org/classes/ActiveSupport/EventReporter.html#class-ActiveSupport::EventReporter-label-Subscribers
[Filtered Subscriptions]: https://api.rubyonrails.org/classes/ActiveSupport/EventReporter.html#class-ActiveSupport::EventReporter-label-Filtered+Subscriptions
[Tags]: https://api.rubyonrails.org/classes/ActiveSupport/EventReporter.html#class-ActiveSupport::EventReporter-label-Tags
[Context Store]: https://api.rubyonrails.org/classes/ActiveSupport/EventReporter.html#class-ActiveSupport::EventReporter-label-Context+Store

[`config.filter_parameters`]: https://guides.rubyonrails.org/configuring.html#config-filter-parameters
[`ActiveSupport::ParameterFilter`]: https://api.rubyonrails.org/classes/ActiveSupport/ParameterFilter.html
[`ActiveSupport::Testing::EventReporterAssertions`]: https://api.rubyonrails.org/classes/ActiveSupport/Testing/EventReporterAssertions.html
[`Rails.event.with_debug`]: https://api.rubyonrails.org/classes/ActiveSupport/EventReporter.html#method-i-with_debug

Framework Hooks (Structured Events Emitted by Rails)
----------------------------------------------------

> Rails emits structured events across the framework covering controllers, jobs, database activity, caching, mailing, streaming, and more. Below you can find a list of these events and their payloads.

Reviewer: Is it worth mentioning which of these events are debug-only? Maybe with a note like:
> NOTE: Events marked with (debug) are only emitted when Rails.event.with_debug is active. See [Debug Events](#debug-events) for details.

| `:filter` | Filter that halted the action |

#### `action_controller.unpermitted_parameters`

| Key                 | Value                                                                   |
| ------------------- | ----------------------------------------------------------------------- |
| `:controller`       | The controller name                                                     |
| `:action`           | The action                                                              |
| `:unpermitted_keys` | The unpermitted keys                                                    |
| `:context`          | Hash with the following keys: `:controller`, `:action`, `:params`       |

#### `action_controller.rescue_from_handled`

| Key                    | Value                          |
| ---------------------- | ------------------------------ |
| `:exception_class`     | The class of the exception     |
| `:exception_message`   | The message of the exception   |
| `:exception_backtrace` | The backtrace of the exception |

### Action Controller: Caching

#### `action_controller.fragment_cache`

| Key            | Value            |
| -------------- | ---------------- |

> | `:method`      | Cache method (e.g. `write_fragment`, `read_fragment`, `exist_fragment?`, `expire_fragment` |

Reviewer: ```suggestion
| `:method`      | Cache method (e.g. `write_fragment`, `read_fragment`, `exist_fragment?`, `expire_fragment`) |
```

| `:key`         | The complete key |
| `:duration_ms` | Duration in ms   |

### Action Dispatch

#### `action_dispatch.redirect`

| Key            | Value                                     |
| -------------- | ----------------------------------------- |
| `:location`    | URL to redirect to                        |
| `:status`      | HTTP response code                        |

> | `:status_name` | HTTP response code                        |

Reviewer: ```suggestion
| `:status_name` | Human-readable HTTP status name (e.g. "OK", "Not Found")   |
```

| `:identifier`   | Full path to template                    |
| `:duration_ms`  | Duration in ms                           |
| `:gc_ms`        | Amount spent in garbage collection in ms |

#### `action_view.render_start`

| Key             | Value                                    |
| --------------- | ---------------------------------------- |
| `:is_layout`    | If the template is a layout              |
| `:identifier`   | Full path to template                    |
| `:layout`       | Applicable layout                        |

### Active Job

#### `active_job.enqueued`

| Key             | Value                        |
| --------------- | ---------------------------- |
| `:job_class`    | Job class name               |
| `:job_id`       | Job ID                       |
| `:queue`        | Queue name                   |
| `:adapter`        | QueueAdapter object processing the job |
| `:aborted`        | Whether this job was aborted |
| `:exception_class`   | The class of the exception if present |
| `:exception_message` | The message of the exception if present |
| `:arguments`         | Arguments passed to the job if `log_arguments` is enabled |

> #### `active_job.enqueue_at`

Reviewer: ```suggestion
#### `active_job.enqueued_at`
```

| `:exception_message` | The message of the exception |

#### `active_job.interrupt`

| Key            | Value                        |
| -------------- | ---------------------------- |
| `:job_class`   | Job class name               |
| `:job_id`      | Job ID                       |
| `:description` | Description of the interrupt |
| `:reason`      | Reason for the interrupt     |

#### `active_job.resume`

| Key            | Value                        |
| -------------- | ---------------------------- |
| `:job_class`   | Job class name               |
| `:job_id`      | Job ID                       |
| `:description` | Description of the interrupt |

#### `active_job.step_skipped`

| Key          | Value                          |
| ------------ | ------------------------------ |
| `:job_class` | Job class name                 |
| `:job_id`    | Job ID                         |
| `:step`      | Name of the step being skipped |

#### `active_job.step_started`

| Key          | Value                            |
| ------------ | -------------------------------- |
| `:job_class` | Job class name                   |
| `:job_id`    | Job ID                           |

> | `:step`      | Name of the step being skipped   |

Reviewer: ```suggestion
| `:step`      | Name of the step being started   |
```

| `:cursor`    | Cursor at which the step resumes |
| `:resumed`   | Whether the step is resuming     |

#### `active_job.step`

| Key                  | Value                            |
| -------------------- | -------------------------------- |
| `:job_class`         | Job class name                   |
| `:job_id`            | Job ID                           |

> | `:step`              | Name of the step being skipped   |

Reviewer: ```suggestion
| `:step`              | Name of the step                 |
```

-------------------------------------------------------------

Introduction to Structured Events
---------------------------------

Structured events are machine-friendly, semantically consistent records describing something that occurred in your application. They are designed for production observability, fast search, analytics platforms, and developer-facing diagnostic tools.

Each event contains:

* a **name**
* a **payload** (hash or event object)
* **timestamp** (nanosecond precision)
* **source location**
* **tags**
* **context**

Rails automatically attaches timestamps, source location, and context so developers only need to provide a name and payload.

### Why Structured Events?

Rails already emits unstructured logs. They are ideal for development but become difficult and expensive to use in production:

* logs from many threads and processes interleave
* no consistent schema
* specialized knowledge needed to query
* indexing costs grow with volume
* onboarding requires understanding bespoke log formats

Structured events fix these problems:

* consistent shape
* semantic naming
* predictable payload keys
* cheap to index in structured stores

> * easy to query

Reviewer 2: ```suggestion
* are easy to query
```

* compatible with any O11Y backend

They also unify **business events** and **developer observability events** behind a single API.

Reporting Events
----------------

The canonical entry point is:

```ruby
Rails.event.notify("event.name", { id: 123 })
```

This attaches metadata and forwards the event to any configured subscribers.

For full method signatures and advanced options (debug, caller depth, filtering rules), see the [EventReporter API documentation][].

> ### Event names

Reviewer 3: ```suggestion
### Event Names
```

Use namespaced identifiers:

```
"controller.request_started"
"auth.user_created"
"graphql.query_resolved"
```

Names should reflect *what happened*, not how or why.

> ### Event payloads

Reviewer 3: ```suggestion
### Event Payloads
```

Payloads may be:

* simple Ruby hashes

> * domain-specific event objects that define their own schema

Reviewer 3: It would be nice if we can show example payloads of these as well.

````suggestion
Payloads can simple Ruby hashes:

```ruby
Rails.event.notify("user_created", { id: 123 })
# Emits event:
#  {
#    name: "user_created",
#    payload: { id: 123 },
#    timestamp: 1738964843208679035,
#    source_location: { filepath: "path/to/file.rb", lineno: 123, label: "UserService#create" }
#  }
```

Or domain-specific event objects that define their own schema:

```ruby
class UserCreatedEvent
  def initialize(id:, name:)
    @id = id
    @name = name
  end

  def serialize
    {
      id: @id,
      name: @name
    }
  end
end

Rails.event.notify(UserCreatedEvent.new(id: 123, name: "John Doe"))
# Emits event:
#  {
#    name: "UserCreatedEvent",
#    payload: #<UserCreatedEvent:0x111>,
#    timestamp: 1738964843208679035,
#    source_location: { filepath: "path/to/file.rb", lineno: 123, label: "UserService#create" }
#  }
```
````

Rails merges any additional keyword arguments into the payload hash.

> For details on object payloads and automatic key normalization, see:

Reviewer 2: ```suggestion
Further details on object payloads and automatic key normalization can be found in the [Event Objects][] section of the API documentation.
```

* **[Event Objects][]**

### Debug Events

Debug events are emitted only when debug mode is enabled:

```ruby
Rails.event.debug("cache.evicted", size: 1024)
```

Use these for high-volume diagnostic telemetry.

For details:

* **[Debug Events][]**

Subscribers
-----------

A subscriber receives events and is responsible for exporting them to a destination such as:

* a log file
* a streaming platform
* a metrics or analytics system
* a SaaS observability backend

Example:

```ruby
class JSONSubscriber
  def emit(event)
    LogExporter.write(JSON.generate(event))
  end
end

Rails.event.subscribe(JSONSubscriber.new)
```

Subscribers may also use filter procs to receive only certain events.

See:

* **[Subscribers][]**
* **[Filtered Subscriptions][]**

Tags
----

Tags annotate an event with domain-specific labels:

```ruby
Rails.event.tagged("graphql") do
  Rails.event.notify("user.created", id: 123)
end
```

Resulting event:

```json
"tags": {
  "graphql": true
}
```

Tags are stack-based and temporary.

> See:

Reviewer 2: ```suggestion
Further details on tags can be found in the [Tags][] section of the API documentation.
```

* **[Tags][]**

Context
-------

Context expresses execution-scoped metadata, such as request identifiers or job metadata. Context is automatically reset between requests and jobs.

```ruby
Rails.event.set_context(request_id: request.uuid)

Rails.event.notify("checkout.succeeded", id: order.id)
```

Resulting event:

```json
"context": {
  "request_id": "abcd123"
}
```

> For custom stores and behavior:

Reviewer 2: ```suggestion
Further details on the context store can be found in the [Context Store][] section of the API documentation.
```

* **[Context Store][]**

Relationship to ActiveSupport::Notifications
--------------------------------------------

Although both systems relate to what happens inside the application, they solve different problems.

### ActiveSupport::Notifications

* an *instrumentation* API
* measures *duration*, *allocations*, *start/finish*
* publishers wrap blocks of code
* subscribers react synchronously

### Structured Events (Event Reporter)

* a *reporting* API
* describes *what happened*, semantically
* payloads are structured
* metadata is attached automatically
* built for observability systems
* subscribers export data externally
* designed for high cardinality, low-cost search

> ### How They Work Together

Reviewer 2: Maybe worth making this a bit more descriptive in case anyone is linked to it in isolation?

```suggestion
### How ActiveSupport::Notifications and Structured Events Work Together
```

`ActiveSupport::StructuredEventSubscriber` bridges the two systems:

1. listens to Notifications (e.g., `"process_action.action_controller"`)
2. transforms them
3. emits structured events via `Rails.event.notify`

Framework components such as Action Controller, Active Record, Active Job, and others use structured subscribers to convert internal instrumentation into structured events.

> Structured Event Subscribers

Author: That's a good point, I guess no we shouldn't document private API.

> Structured Event Subscribers

Author: @adrianna-chang-shopify One small note, the internal subscribers are private but the top-level `StructuredEventSubscriber` is not -- there is a use-case for using that to neatly emit your custom ASN to event reporter, but I think it's too specific for this guide. Just wanted to dump there here where I can resolve it hehe.

----------------------------

A structured subscriber is a specialized Notifications subscriber that turns instrumentation events into structured events.

Example:

```ruby
class StructuredControllerSubscriber < ActiveSupport::StructuredEventSubscriber
  attach_to :action_controller

  def start_processing(event)
    emit_event("controller.request_started",
      controller: event.payload[:controller],
      action: event.payload[:action],
      format: event.payload[:format]
    )
  end
end
```

> Key behaviors:

Reviewer 2: ```suggestion
A structured subscriber:
```

This guide explains the concepts, design, and usage patterns behind structured events, and how Rails integrates them with framework instrumentation.

After reading this guide, you will understand:

* What structured events are
* When to use them
* How Rails reports events
* How subscribers consume and export events
* How tags and context work
* How structured subscribers connect Notifications -> Events
* Where to find the authoritative API definitions
* The structured events emitted by Rails (framework hooks)

-------------------------------------------------------------

Introduction to Structured Events
---------------------------------

Structured events are machine-friendly, semantically consistent records describing something that occurred in your application. They are designed for production observability, fast search, analytics platforms, and developer-facing diagnostic tools.

Each event contains:

* a **name**
* a **payload** (hash or event object)
* **timestamp** (nanosecond precision)
* **source location**
* **tags**
* **context**

> Rails automatically attaches timestamps, source location, and context so developers only need to provide a name and payload.

Reviewer: Nit, but technically Rails doesn't supply the context -- the user supplies the context, and Rails makes sure it propagates to all structured logs in the stack.

### Why Structured Events?

Rails already emits unstructured logs. They are ideal for development but become difficult and expensive to use in production. Some of the challenges of logs are as follows:

* logs from many threads and processes interleave
* there is no consistent schema
* specialized knowledge is needed to query the logs
* indexing costs grow with volume
* onboarding requires an understanding of bespoke log formats

Structured events fix these problems as they:

* have a consistent shape
* have semantic naming
* have predictable payload keys
* are cheap to index in structured stores
* are easy to query

> * are compatible with any Obeservability backend

Reviewer: ```suggestion
* are compatible with any Observability backend
```

Although maybe I'd say "integrate easily with Observability backends" instead? ie. in the case of event objects, subscribers may need to serialize them into a shape that ends up being compatible with the observability backend being used.
